In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
from pydataset import data

# 1. Attendance Data

### Load the attendance.csv file. 

In [29]:
attendance = pd.read_csv('untidy-data/attendance.csv')
attendance

Unnamed: 0 2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05  \
0      Sally          P          T          T          H          P   
1       Jane          A          P          T          T          T   
2      Billy          A          T          A          A          H   
3       John          P          T          H          P          P   

  2018-01-06 2018-01-07 2018-01-08  
0          A          T          T  
1          T          A          T  
2          T          P          T  
3          T          P          P

### Calculate an attendance percentage for each student. One half day is worth 50% of a full day, and 10 tardies is equal to one absence.

In [30]:
attendance = attendance.melt(id_vars =  ['Unnamed: 0'])

In [31]:
attendance.head()

Unnamed: 0    variable value
0      Sally  2018-01-01     P
1       Jane  2018-01-01     A
2      Billy  2018-01-01     A
3       John  2018-01-01     P
4      Sally  2018-01-02     T

In [33]:
attendance.columns = ['name', 'date', 'attendance_type']
attendance.head()

name        date attendance_type
0  Sally  2018-01-01               P
1   Jane  2018-01-01               A
2  Billy  2018-01-01               A
3   John  2018-01-01               P
4  Sally  2018-01-02               T

In [34]:
attendance_weights = {'P': 1, 'T': 0.9, 'H': 0.5, 'A': 0}

In [37]:
attendance = attendance.replace({'attendance_type': attendance_weights})
attendance.head()

name        date  attendance_type
0  Sally  2018-01-01              1.0
1   Jane  2018-01-01              0.0
2  Billy  2018-01-01              0.0
3   John  2018-01-01              1.0
4  Sally  2018-01-02              0.9

In [38]:
attendance.groupby('name').attendance_type.mean()

name
Billy    0.5250
Jane     0.6875
John     0.9125
Sally    0.7625
Name: attendance_type, dtype: float64

# 2. Coffee Levels

### Read the coffee_levels.csv file.

In [48]:
coffee_levels = pd.read_csv('untidy-data/coffee_levels.csv')

In [49]:
coffee_levels

hour coffee_carafe  coffee_amount
0      8             x       0.816164
1      9             x       0.451018
2     10             x       0.843279
3     11             x       0.335533
4     12             x       0.898291
5     13             x       0.310711
6     14             x       0.507288
7     15             x       0.215043
8     16             x       0.183891
9     17             x       0.391560
10     8             y       0.189297
11     9             y       0.521502
12    10             y       0.023163
13    11             y       0.235529
14    12             y       0.017009
15    13             y       0.997464
16    14             y       0.058361
17    15             y       0.144644
18    16             y       0.544676
19    17             y       0.594126
20     8             z       0.999264
21     9             z       0.915990
22    10             z       0.144928
23    11             z       0.311495
24    12             z       0.771947
25    13             z       0.398520
26    14             z       0.864464
27    15             z       0.436364
28    16             z       0.280621
29    17             z       0.436677

### Transform the data so that each carafe is in it's own column.

In [50]:
coffee_levels = coffee_levels.pivot_table(index = 'hour',
                   columns = 'coffee_carafe',
                   values = 'coffee_amount').reset_index()

In [51]:
coffee_levels

coffee_carafe  hour         x         y         z
0                 8  0.816164  0.189297  0.999264
1                 9  0.451018  0.521502  0.915990
2                10  0.843279  0.023163  0.144928
3                11  0.335533  0.235529  0.311495
4                12  0.898291  0.017009  0.771947
5                13  0.310711  0.997464  0.398520
6                14  0.507288  0.058361  0.864464
7                15  0.215043  0.144644  0.436364
8                16  0.183891  0.544676  0.280621
9                17  0.391560  0.594126  0.436677

### Is this the best shape for the data?

The second shape is better for the data. If our goal is to know the coffee levels in each carafe at different hours of the data, the second graph is much easier to understand. The side by side columns allow us to compare coffee levels in each carafe at different times of day. Such a comparison is difficult to discern in from the first graph.

# 3. Cake Recipes

### Read the cake_recipes.csv data.

In [90]:
cake_recipes = pd.read_csv('untidy-data/cake_recipes.csv')
cake_recipes

recipe:position        225        250        275        300
0        a:bottom  61.738655  53.912627  74.414730  98.786784
1           a:top  51.709751  52.009735  68.576858  50.228470
2        b:bottom  57.095320  61.904369  61.196980  99.248541
3           b:top  82.455004  95.224151  98.594881  58.169349
4        c:bottom  96.470207  52.001358  92.893227  65.473084
5           c:top  71.306308  82.795477  92.098049  53.960273
6        d:bottom  52.799753  58.670419  51.747686  56.183110
7           d:top  96.873178  76.101363  59.571620  50.971626

This data set contains cake tastiness scores for combinations of different recipes, oven rack positions, and oven temperatures.

### Tidy the data as necessary.

In [91]:
cake_recipes[['recipe', 'oven_rack_position']]  = cake_recipes['recipe:position'].str.split(':', expand = True)
cake_recipes

recipe:position        225        250        275        300 recipe  \
0        a:bottom  61.738655  53.912627  74.414730  98.786784      a   
1           a:top  51.709751  52.009735  68.576858  50.228470      a   
2        b:bottom  57.095320  61.904369  61.196980  99.248541      b   
3           b:top  82.455004  95.224151  98.594881  58.169349      b   
4        c:bottom  96.470207  52.001358  92.893227  65.473084      c   
5           c:top  71.306308  82.795477  92.098049  53.960273      c   
6        d:bottom  52.799753  58.670419  51.747686  56.183110      d   
7           d:top  96.873178  76.101363  59.571620  50.971626      d   

  oven_rack_position  
0             bottom  
1                top  
2             bottom  
3                top  
4             bottom  
5                top  
6             bottom  
7                top

In [92]:
cake_recipes = cake_recipes.drop("recipe:position", axis =1)
cake_recipes

225        250        275        300 recipe oven_rack_position
0  61.738655  53.912627  74.414730  98.786784      a             bottom
1  51.709751  52.009735  68.576858  50.228470      a                top
2  57.095320  61.904369  61.196980  99.248541      b             bottom
3  82.455004  95.224151  98.594881  58.169349      b                top
4  96.470207  52.001358  92.893227  65.473084      c             bottom
5  71.306308  82.795477  92.098049  53.960273      c                top
6  52.799753  58.670419  51.747686  56.183110      d             bottom
7  96.873178  76.101363  59.571620  50.971626      d                top

In [95]:
cake_recipes = cake_recipes.melt(id_vars =  ['recipe', 'oven_rack_position'], var_name = 'oven_temperature', value_name = 'score')
cake_recipes

recipe oven_rack_position oven_temperature      score
0       a             bottom              225  61.738655
1       a                top              225  51.709751
2       b             bottom              225  57.095320
3       b                top              225  82.455004
4       c             bottom              225  96.470207
5       c                top              225  71.306308
6       d             bottom              225  52.799753
7       d                top              225  96.873178
8       a             bottom              250  53.912627
9       a                top              250  52.009735
10      b             bottom              250  61.904369
11      b                top              250  95.224151
12      c             bottom              250  52.001358
13      c                top              250  82.795477
14      d             bottom              250  58.670419
15      d                top              250  76.101363
16      a             bottom              275  74.414730
17      a                top              275  68.576858
18      b             bottom              275  61.196980
19      b                top              275  98.594881
20      c             bottom              275  92.893227
21      c                top              275  92.098049
22      d             bottom              275  51.747686
23      d                top              275  59.571620
24      a             bottom              300  98.786784
25      a                top              300  50.228470
26      b             bottom              300  99.248541
27      b                top              300  58.169349
28      c             bottom              300  65.473084
29      c                top              300  53.960273
30      d             bottom              300  56.183110
31      d                top              300  50.971626

In [96]:
cake_recipes = cake_recipes.sort_values(by=['recipe', 'oven_rack_position'])
cake_recipes

recipe oven_rack_position oven_temperature      score
0       a             bottom              225  61.738655
8       a             bottom              250  53.912627
16      a             bottom              275  74.414730
24      a             bottom              300  98.786784
1       a                top              225  51.709751
9       a                top              250  52.009735
17      a                top              275  68.576858
25      a                top              300  50.228470
2       b             bottom              225  57.095320
10      b             bottom              250  61.904369
18      b             bottom              275  61.196980
26      b             bottom              300  99.248541
3       b                top              225  82.455004
11      b                top              250  95.224151
19      b                top              275  98.594881
27      b                top              300  58.169349
4       c             bottom              225  96.470207
12      c             bottom              250  52.001358
20      c             bottom              275  92.893227
28      c             bottom              300  65.473084
5       c                top              225  71.306308
13      c                top              250  82.795477
21      c                top              275  92.098049
29      c                top              300  53.960273
6       d             bottom              225  52.799753
14      d             bottom              250  58.670419
22      d             bottom              275  51.747686
30      d             bottom              300  56.183110
7       d                top              225  96.873178
15      d                top              250  76.101363
23      d                top              275  59.571620
31      d                top              300  50.971626

### Which recipe, on average, is the best? 

In [99]:
cake_recipes.groupby('recipe').score.mean()

recipe
a    63.922201
b    76.736074
c    75.874748
d    62.864844
Name: score, dtype: float64

In [103]:
cake_recipes.groupby('recipe').score.mean().max()

76.73607424644436

As we can see, the "b" recipe has the highest average rating at approximately 76.74.

### Which oven temperature, on average, produces the best results? 

In [104]:
cake_recipes.groupby('oven_temperature').score.mean()

oven_temperature
225    71.306022
250    66.577437
275    74.886754
300    66.627655
Name: score, dtype: float64

In [105]:
cake_recipes.groupby('oven_temperature').score.mean().max()

74.8867538664983

As we can see, an oven temperature of 275 produces the highest average rating, with an average score of approximately 74.89.

### Which combination of recipe, rack position, and temperature gives the best result?

In [108]:
highest_rated = cake_recipes.score == cake_recipes.score.max()

In [109]:
cake_recipes[highest_rated]

recipe oven_rack_position oven_temperature      score
26      b             bottom              300  99.248541

Recipe "b", at the bottom oven rack position, with an oven temperature of 300 produces the highest score, with a rating of approximately 99.25.